File: src/main/java/com/gitlab/daring/image/sandbox/ImageSandbox.java
package com.gitlab.daring.image.sandbox;

import com.gitlab.daring.image.common.BaseComponent;
import com.gitlab.daring.image.swing.BaseFrame;

import static com.gitlab.daring.image.MainContext.mainContext;
import static com.gitlab.daring.image.sandbox.SandboxUtils.saveCompConfig;

public class ImageSandbox extends BaseComponent implements AutoCloseable {

	static final String ConfigPath = "gmv.ImageSandbox";

	final MainPanel mp = new MainPanel(this);
	final ScriptExecutor scriptExecutor = new ScriptExecutor(this);

	public ImageSandbox() {
		super(ConfigPath);
		mp.applyEvent.onFire(this::apply);
		mp.changeEvent.onFire(this::executeScript);
		executeScript();
	}

	void showFrame() {
		BaseFrame frame = new BaseFrame("ImageSandbox", mp);
		frame.addCloseListener(this::close);
		frame.show(800, 600);
	}

	void apply() {
		executeScript();
		saveConfig();
	}

	void executeScript() {
		scriptExecutor.executeAsync();
	}

	void saveConfig() {
		saveCompConfig(ConfigPath, m -> {
			m.put("script", mp.script.getText());
			m.put("files", mp.filesParam.v);
		});
	}

	@Override
	public void close() {
		scriptExecutor.close();
		mainContext().close();
	}

	public static void main(String[] args) {
		new ImageSandbox().showFrame();
	}

}

File: src/main/kotlin/com/gitlab/daring/image/sandbox/ImageSandbox.kt
package com.gitlab.daring.image.sandbox

import com.gitlab.daring.image.MainContext
import com.gitlab.daring.image.config.ConfigUtils.defaultConfig
import com.gitlab.daring.image.swing.BaseFrame

internal class ImageSandbox : AutoCloseable {

    val ConfigPath = "isb.ImageSandbox"
    val config = defaultConfig.getConfig(ConfigPath)

    val mp = MainPanel(this)
    val scriptExecutor = ScriptExecutor(this)

    init {
        mp.applyEvent.onFire(this::apply)
        mp.changeEvent.onFire(this::executeScript)
    }

    fun showFrame() {
        val frame = BaseFrame("ImageSandbox", mp)
        frame.addCloseListener(this::close)
        frame.show(800, 600)
    }

    fun apply() {
        executeScript()
        mp.saveConfig(ConfigPath)
    }

    fun executeScript() {
        scriptExecutor.executeAsync()
    }

    override fun close() {
        scriptExecutor.close()
        MainContext.close()
    }

    companion object {

        @JvmStatic
        fun main(args: Array<String>) {
            ImageSandbox().showFrame()
        }

    }

}
